File: frontend/src/services/answerService.ts
import { gql } from "@apollo/client/core";
import { apolloClient } from "../apollo-client.ts";

export interface Answer {
  id: string;
  questionId: string;
  answer: string;
  isCorrect: boolean;
  image?: string;
  createdAt: string;
}

export const answerService = {
  getAnswer: async (id: string) => {
    const { data } = await apolloClient.query({
      query: gql`
        query GetAnswer($id: ID!) {
          answer(id: $id) {
            id
            questionId
            answer
            isCorrect
            image
            createdAt
          }
        }
      `,
      variables: { id },
    });
    return data.answer;
  },

  getAnswers: async (questionId: string) => {
    const { data } = await apolloClient.query({
      query: gql`
        query GetAnswers($questionId: ID!) {
          answers(questionId: $questionId) {
            id
            questionId
            answer
            isCorrect
            image
            createdAt
          }
        }
      `,
      variables: { questionId },
    });
    return data.answers;
  },

  createAnswer: async (
    questionId: string,
    answer: string,
    isCorrect: boolean,
    image?: string
  ) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation CreateAnswer(
          $questionId: ID!
          $answer: String!
          $isCorrect: Boolean!
          $image: String
        ) {
          createAnswer(
            questionId: $questionId
            answer: $answer
            isCorrect: $isCorrect
            image: $image
          ) {
            id
            questionId
            answer
            isCorrect
            image
            createdAt
          }
        }
      `,
      variables: { questionId, answer, isCorrect, image },
    });
    return data.createAnswer;
  },

  updateAnswer: async (
    id: string,
    answer?: string,
    isCorrect?: boolean,
    image?: string
  ) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation UpdateAnswer(
          $id: ID!
          $answer: String
          $isCorrect: Boolean
          $image: String
        ) {
          updateAnswer(
            id: $id
            answer: $answer
            isCorrect: $isCorrect
            image: $image
          ) {
            id
            questionId
            answer
            isCorrect
            image
            createdAt
          }
        }
      `,
      variables: { id, answer, isCorrect, image },
    });
    return data.updateAnswer;
  },

  deleteAnswer: async (id: string) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation DeleteAnswer($id: ID!) {
          deleteAnswer(id: $id)
        }
      `,
      variables: { id },
    });
    return data.deleteAnswer;
  },
};

File: flutter/lib/data/services/answer_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:graphql_flutter/graphql_flutter.dart';
import '../lib/models/answer_model.dart';
import 'auth_service.dart';

class AnswerService {
  final GraphQLClient _client = AuthService().client;
  final secureStorage = const FlutterSecureStorage();

  Future<Context?> _getHeadersContext() async {
    final String? token = await secureStorage.read(key: "jwt");

    if (token != null && token.isNotEmpty) {
      return Context.fromList([
        HttpLinkHeaders(headers: {'Authorization': 'Bearer $token'}),
      ]);
    }
    return null;
  }

  Future<List<Answer>> getAnswers({required String questionId}) async {
    final String query = """
      query GetAnswers(\$questionId: ID!) {
        answers(questionId: \$questionId) {
          id
          questionId
          answer
          isCorrect
          image
          createdAt
        }
      }
    """;

    final QueryOptions options = QueryOptions(
      document: gql(query),
      variables: {'questionId': questionId},
      fetchPolicy: FetchPolicy.networkOnly,
      context: await _getHeadersContext(),
    );

    final QueryResult result = await _client.query(options);

    if (result.hasException) {
      print(result.exception.toString());
      throw Exception('Failed to load answers');
    }

    final List<dynamic> answersJson = result.data?['answers'] ?? [];
    return answersJson.map((json) => Answer.fromJson(json)).toList();
  }
}
